This patch is based on the commit described below, taken from the upstream
time64 pull request, adapted for use in the Debian package by Peter Michael
Green.

commit fc5c42dd8d00acd651a53950d44b6284a00979b0
Author: Ola x Nilsson <olani@axis.com>
Date:   Fri Mar 17 15:20:29 2023 +0100

    gnu: Move struct timex_t to gnu/b* and support _TIME_BITS=64

Index: libc/src/unix/linux_like/linux/gnu/b32/mod.rs
===================================================================
--- libc.orig/src/unix/linux_like/linux/gnu/b32/mod.rs
+++ libc/src/unix/linux_like/linux/gnu/b32/mod.rs
@@ -1,6 +1,7 @@
 //! 32-bit specific definitions for linux-like values
 
 use crate::prelude::*;
+use crate::ntptimeval;
 use crate::pthread_mutex_t;
 
 pub type c_long = i32;
@@ -18,6 +19,7 @@ pub type fsblkcnt64_t = u64;
 pub type fsfilcnt64_t = u64;
 pub type __syscall_ulong_t = c_ulong;
 
+pub type __syscall_slong_t = c_long;
 cfg_if! {
     if #[cfg(target_arch = "riscv32")] {
         pub type time_t = i64;
@@ -170,6 +172,92 @@ s! {
         __glibc_reserved3: crate::__syscall_ulong_t,
         __glibc_reserved4: crate::__syscall_ulong_t,
     }
+
+    pub struct timex {
+        pub modes: c_uint,
+
+        #[cfg(not(gnu_time64_abi))]
+        pub offset: crate::__syscall_slong_t,
+        #[cfg(gnu_time64_abi)]
+        pub __unused_pad1: i32,
+        #[cfg(gnu_time64_abi)]
+        pub offset: c_longlong,
+        #[cfg(not(gnu_time64_abi))]
+        pub freq: crate::__syscall_slong_t,
+        #[cfg(gnu_time64_abi)]
+        pub freq: c_longlong,
+        #[cfg(not(gnu_time64_abi))]
+        pub maxerror: crate::__syscall_slong_t,
+        #[cfg(gnu_time64_abi)]
+        pub maxerror: c_longlong,
+        #[cfg(not(gnu_time64_abi))]
+        pub esterror: crate::__syscall_slong_t,
+        #[cfg(gnu_time64_abi)]
+        pub esterror: c_longlong,
+        pub status: c_int,
+        #[cfg(not(gnu_time64_abi))]
+        pub constant: crate::__syscall_slong_t,
+        #[cfg(gnu_time64_abi)]
+        pub __unused_pad2: i32,
+        #[cfg(gnu_time64_abi)]
+        pub constant: c_longlong,
+        #[cfg(not(gnu_time64_abi))]
+        pub precision: crate::__syscall_slong_t,
+        #[cfg(gnu_time64_abi)]
+        pub precision: c_longlong,
+        #[cfg(not(gnu_time64_abi))]
+        pub tolerance: crate::__syscall_slong_t,
+        #[cfg(gnu_time64_abi)]
+        pub tolerance: c_longlong,
+        pub time: crate::timeval,
+        #[cfg(not(gnu_time64_abi))]
+        pub tick: crate::__syscall_slong_t,
+        #[cfg(gnu_time64_abi)]
+        pub tick: c_longlong,
+        #[cfg(not(gnu_time64_abi))]
+        pub ppsfreq: crate::__syscall_slong_t,
+        #[cfg(gnu_time64_abi)]
+        pub ppsfreq: c_longlong,
+        #[cfg(not(gnu_time64_abi))]
+        pub jitter: crate::__syscall_slong_t,
+        #[cfg(gnu_time64_abi)]
+        pub jitter: c_longlong,
+        pub shift: c_int,
+        #[cfg(not(gnu_time64_abi))]
+        pub stabil: crate::__syscall_slong_t,
+        #[cfg(gnu_time64_abi)]
+        pub __unused_pad3: i32,
+        #[cfg(gnu_time64_abi)]
+        pub stabil: c_longlong,
+        #[cfg(not(gnu_time64_abi))]
+        pub jitcnt: crate::__syscall_slong_t,
+        #[cfg(gnu_time64_abi)]
+        pub jitcnt: c_longlong,
+        #[cfg(not(gnu_time64_abi))]
+        pub calcnt: crate::__syscall_slong_t,
+        #[cfg(gnu_time64_abi)]
+        pub calcnt: c_longlong,
+        #[cfg(not(gnu_time64_abi))]
+        pub errcnt: crate::__syscall_slong_t,
+        #[cfg(gnu_time64_abi)]
+        pub errcnt: c_longlong,
+        #[cfg(not(gnu_time64_abi))]
+        pub stbcnt: crate::__syscall_slong_t,
+        #[cfg(gnu_time64_abi)]
+        pub stbcnt: c_longlong,
+        pub tai: c_int,
+        pub __unused1: i32,
+        pub __unused2: i32,
+        pub __unused3: i32,
+        pub __unused4: i32,
+        pub __unused5: i32,
+        pub __unused6: i32,
+        pub __unused7: i32,
+        pub __unused8: i32,
+        pub __unused9: i32,
+        pub __unused10: i32,
+        pub __unused11: i32,
+    }
 }
 
 pub const POSIX_FADV_DONTNEED: c_int = 4;
Index: libc/src/unix/linux_like/linux/gnu/b64/mod.rs
===================================================================
--- libc.orig/src/unix/linux_like/linux/gnu/b64/mod.rs
+++ libc/src/unix/linux_like/linux/gnu/b64/mod.rs
@@ -2,6 +2,8 @@
 
 use crate::prelude::*;
 
+use crate::ntptimeval;
+
 pub type ino_t = u64;
 pub type off_t = i64;
 pub type blkcnt_t = i64;
@@ -11,12 +13,17 @@ pub type msglen_t = u64;
 pub type fsblkcnt_t = u64;
 pub type fsfilcnt_t = u64;
 pub type rlim_t = u64;
-#[cfg(all(target_arch = "x86_64", target_pointer_width = "32"))]
-pub type __syscall_ulong_t = c_ulonglong;
-#[cfg(not(all(target_arch = "x86_64", target_pointer_width = "32")))]
-pub type __syscall_ulong_t = c_ulong;
 
 cfg_if! {
+    if #[cfg(all(target_arch = "x86_64", target_pointer_width = "32"))] {
+        pub type __syscall_ulong_t = c_ulonglong;
+        pub type __syscall_slong_t = c_longlong;
+    } else {
+        pub type __syscall_ulong_t = c_ulong;
+        pub type __syscall_slong_t = c_long;
+    }
+}
+cfg_if! {
     if #[cfg(all(target_arch = "aarch64", target_pointer_width = "32"))] {
         pub type clock_t = i32;
         pub type time_t = i32;
@@ -95,6 +102,46 @@ s! {
         __glibc_reserved3: crate::__syscall_ulong_t,
         __glibc_reserved4: crate::__syscall_ulong_t,
     }
+
+    pub struct timex {
+        pub modes: c_uint,
+        #[cfg(all(target_arch = "x86_64", target_pointer_width = "32"))]
+        pub __unused_pad1: i32,
+        pub offset: crate::__syscall_slong_t,
+        pub freq: crate::__syscall_slong_t,
+        pub maxerror: crate::__syscall_slong_t,
+        pub esterror: crate::__syscall_slong_t,
+        pub status: c_int,
+        #[cfg(all(target_arch = "x86_64", target_pointer_width = "32"))]
+        pub __unused_pad2: i32,
+        pub constant: crate::__syscall_slong_t,
+        pub precision: crate::__syscall_slong_t,
+        pub tolerance: crate::__syscall_slong_t,
+        pub time: crate::timeval,
+        pub tick: crate::__syscall_slong_t,
+        pub ppsfreq: crate::__syscall_slong_t,
+        pub jitter: crate::__syscall_slong_t,
+        pub shift: c_int,
+        #[cfg(all(target_arch = "x86_64", target_pointer_width = "32"))]
+        pub __unused_pad3: i32,
+        pub stabil: crate::__syscall_slong_t,
+        pub jitcnt: crate::__syscall_slong_t,
+        pub calcnt: crate::__syscall_slong_t,
+        pub errcnt: crate::__syscall_slong_t,
+        pub stbcnt: crate::__syscall_slong_t,
+        pub tai: c_int,
+        pub __unused1: i32,
+        pub __unused2: i32,
+        pub __unused3: i32,
+        pub __unused4: i32,
+        pub __unused5: i32,
+        pub __unused6: i32,
+        pub __unused7: i32,
+        pub __unused8: i32,
+        pub __unused9: i32,
+        pub __unused10: i32,
+        pub __unused11: i32,
+    }
 }
 
 pub const __SIZEOF_PTHREAD_RWLOCKATTR_T: usize = 8;
Index: libc/src/unix/linux_like/linux/gnu/mod.rs
===================================================================
--- libc.orig/src/unix/linux_like/linux/gnu/mod.rs
+++ libc/src/unix/linux_like/linux/gnu/mod.rs
@@ -170,85 +170,6 @@ s! {
         pub rt_irtt: c_ushort,
     }
 
-    pub struct timex {
-        pub modes: c_uint,
-        #[cfg(all(target_arch = "x86_64", target_pointer_width = "32"))]
-        pub offset: i64,
-        #[cfg(not(all(target_arch = "x86_64", target_pointer_width = "32")))]
-        pub offset: c_long,
-        #[cfg(all(target_arch = "x86_64", target_pointer_width = "32"))]
-        pub freq: i64,
-        #[cfg(not(all(target_arch = "x86_64", target_pointer_width = "32")))]
-        pub freq: c_long,
-        #[cfg(all(target_arch = "x86_64", target_pointer_width = "32"))]
-        pub maxerror: i64,
-        #[cfg(not(all(target_arch = "x86_64", target_pointer_width = "32")))]
-        pub maxerror: c_long,
-        #[cfg(all(target_arch = "x86_64", target_pointer_width = "32"))]
-        pub esterror: i64,
-        #[cfg(not(all(target_arch = "x86_64", target_pointer_width = "32")))]
-        pub esterror: c_long,
-        pub status: c_int,
-        #[cfg(all(target_arch = "x86_64", target_pointer_width = "32"))]
-        pub constant: i64,
-        #[cfg(not(all(target_arch = "x86_64", target_pointer_width = "32")))]
-        pub constant: c_long,
-        #[cfg(all(target_arch = "x86_64", target_pointer_width = "32"))]
-        pub precision: i64,
-        #[cfg(not(all(target_arch = "x86_64", target_pointer_width = "32")))]
-        pub precision: c_long,
-        #[cfg(all(target_arch = "x86_64", target_pointer_width = "32"))]
-        pub tolerance: i64,
-        #[cfg(not(all(target_arch = "x86_64", target_pointer_width = "32")))]
-        pub tolerance: c_long,
-        pub time: crate::timeval,
-        #[cfg(all(target_arch = "x86_64", target_pointer_width = "32"))]
-        pub tick: i64,
-        #[cfg(not(all(target_arch = "x86_64", target_pointer_width = "32")))]
-        pub tick: c_long,
-        #[cfg(all(target_arch = "x86_64", target_pointer_width = "32"))]
-        pub ppsfreq: i64,
-        #[cfg(not(all(target_arch = "x86_64", target_pointer_width = "32")))]
-        pub ppsfreq: c_long,
-        #[cfg(all(target_arch = "x86_64", target_pointer_width = "32"))]
-        pub jitter: i64,
-        #[cfg(not(all(target_arch = "x86_64", target_pointer_width = "32")))]
-        pub jitter: c_long,
-        pub shift: c_int,
-        #[cfg(all(target_arch = "x86_64", target_pointer_width = "32"))]
-        pub stabil: i64,
-        #[cfg(not(all(target_arch = "x86_64", target_pointer_width = "32")))]
-        pub stabil: c_long,
-        #[cfg(all(target_arch = "x86_64", target_pointer_width = "32"))]
-        pub jitcnt: i64,
-        #[cfg(not(all(target_arch = "x86_64", target_pointer_width = "32")))]
-        pub jitcnt: c_long,
-        #[cfg(all(target_arch = "x86_64", target_pointer_width = "32"))]
-        pub calcnt: i64,
-        #[cfg(not(all(target_arch = "x86_64", target_pointer_width = "32")))]
-        pub calcnt: c_long,
-        #[cfg(all(target_arch = "x86_64", target_pointer_width = "32"))]
-        pub errcnt: i64,
-        #[cfg(not(all(target_arch = "x86_64", target_pointer_width = "32")))]
-        pub errcnt: c_long,
-        #[cfg(all(target_arch = "x86_64", target_pointer_width = "32"))]
-        pub stbcnt: i64,
-        #[cfg(not(all(target_arch = "x86_64", target_pointer_width = "32")))]
-        pub stbcnt: c_long,
-        pub tai: c_int,
-        pub __unused1: i32,
-        pub __unused2: i32,
-        pub __unused3: i32,
-        pub __unused4: i32,
-        pub __unused5: i32,
-        pub __unused6: i32,
-        pub __unused7: i32,
-        pub __unused8: i32,
-        pub __unused9: i32,
-        pub __unused10: i32,
-        pub __unused11: i32,
-    }
-
     pub struct ntptimeval {
         pub time: crate::timeval,
         pub maxerror: c_long,
